federationapi/routing/send.go

```diff
@@ -211,8 +216,27 @@ func (t *txnReq) processEventWithMissingState(e gomatrixserverlib.Event) error {
 		return err
 	}
 	// Check that the event is allowed by the state.
+retryAllowedState:
 	if err := checkAllowedByState(e, state.StateEvents); err != nil {
-		return err
+		switch missing := err.(type) {
+		case gomatrixserverlib.MissingAuthEventError:
+			// An auth event was missing so let's look up that event over federation
+			for _, s := range state.StateEvents {
+				if s.EventID() == missing.AuthEventID {
+					if serr := t.processEventWithMissingState(s, depth+1); serr == nil {
+						// If there was no error retrieving the event from federation then
+						// we assume that it succeeded, so retry the original state check
+						goto retryAllowedState
+					} else {
+						// An error occurred so let's not do anything further
+						return err
```

Comment:
This should be `serr`. It feels weird to be doing the `== nil` case first then this. Why are we doing it this way?

Reply:
I've refactored it now, should be more straight-forward.